### ISSUE TITLE
'Cannot substract NO_VALUE from long' error when substracting null 

### ISSUE THREAD

**Issue by v2belleville** [neo4j/neo4j#13327]
the following query was working fine in 5.9, although there were null values for birthYear. with 5.11, it is now throwing an error 
![image](https://github.com/neo4j/neo4j/assets/8779004/537cc0bc-e726-40fb-98ed-3a38fbfb7e36)

this does work, but I much prefer some sensical default behavior rather than having to write so much code to deal with null values... 
if the new behavior is intended, I'd be interested in reading why this choice has been made...
![image](https://github.com/neo4j/neo4j/assets/8779004/5159cb50-2976-46e0-a915-3b2c25991918)

**Comment by pontusmelke** [comment]
Hi @v2belleville, thanks for reporting this.

I have been trying to reproduce the issue but so far I have not. Do you have anymore details share? Maybe you can run the query with EXPLAIN and attach the profile here? What do you get if you simply do `WITH 42 AS a, NULL AS b RETURN toInteger(a) - toInteger(b) AS res`?

Best regards,
Pontus

**Comment by v2belleville** [comment]
this returns null, shouldn't it be 42 ? no error though...
![image](https://github.com/neo4j/neo4j/assets/8779004/58eb9b9f-8df3-4bf7-95a7-24a1c81fc746)

here is the plan
![plan](https://github.com/neo4j/neo4j/assets/8779004/9588fca1-0532-4617-9422-2bc9bd721f94)

here are the detected types of the properties
![image](https://github.com/neo4j/neo4j/assets/8779004/0d2d7b8f-1884-479c-adbd-6e9a68dca2e6)

**Comment by pontusmelke** [comment]
No it should be `NULL` in that case. `NULL` should be regarded as  "unknown" here, if you add or subtract something to an "unknown" value the result is also "unknown". It shouldn't throw an error though, so far I am not seeing that on my side though but I will continue the investigation.

**Comment by v2belleville** [comment]
we are not substracting `null`, we are substracting `toInteger(null)` and I would expect  `toInteger(null)` to be zero

the error still is puzzling though. good luck in chasing it down ;-)

**Comment by mnd999** [comment]
`toInteger(null)` is defined to be `null`  see the [documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-tointeger) here. If you want to catch nulls, you can use [coalesce](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-coalesce)

**Comment by pontusmelke** [comment]
Yes by the same logic, `NULL` is "unknown" so is `toInteger(NULL)`. Looking at that plan that must be from the variant of the plan that succeeds? The one that explicitly checks if `toInteger` is `NULL`?

Or wait it must be from some different query.  From reading the plan the query where that plan was coming from does not have a `SET` and it has a `RETURN`, which is not true for the initial query you posted?

**Comment by v2belleville** [comment]
not sure what you are asking @pontusmelke 
since `toInteger(null)` is defined as `null` the error thrown is an expected error. (the query worked in 5.9, 5.11 must have been updated )

**Comment by pontusmelke** [comment]
@v2belleville sorry if I wasn't being very clear. There shouldn't be an error here. I have just not been able to reproduce the error, but if you see one it is indeed a bug. The plans you shared cannot be from the initial query

![image](https://github.com/neo4j/neo4j/assets/88000/6a038233-46ff-4e2e-8e6f-cf1fca38829a)

The plan you attached looks like it is for something like

```
MATCH (p:Person)<-[:familyChild]-(f:Family) 
RETURN toInteger(...) - toInteger(...)
```

and that was not where you saw the failure right?

**Comment by v2belleville** [comment]
yes that is what triggers the error but if `toInteger(null)`  returns `null` is it not expected that it throws an error when deadYear is an integer and birthYear is null , meaning it cannot subtract null from an integer ?

**Comment by pontusmelke** [comment]
No `a - b` should evaluate to `NULL` if either `a` or `b` is `NULL`.

**Comment by v2belleville** [comment]
here is the simplified query that triggers the error and the corresponding plan
![image](https://github.com/neo4j/neo4j/assets/8779004/bc869ca3-255d-41e1-bb1f-97b88e2bb3bc)
![plan](https://github.com/neo4j/neo4j/assets/8779004/009a49d2-a771-4b6a-9390-0d4ed4186ca4)

**Comment by v2belleville** [comment]
here are the types returned by `toInteger(p.deathYear)` and `toInteger(p.birthYear)`

![image](https://github.com/neo4j/neo4j/assets/8779004/adf5e8b3-733a-4a56-a8c7-0cdc2a8fac95)

**Comment by pontusmelke** [comment]
Hi again,
I have still not been able reproduce the exception. Do you think you could have a look in debug.log and see if you can find the full stack trace there? (i.e. not just the _Cannot subtract NO_VALUE from `Long`_)

Regards,
Pontus

**Comment by pontusmelke** [comment]
I think I may have discovered something, I can get the error you are seeing if the p.deathYear or p.birthYear is a string that cannot be parsed as an integer. I will provide a fix for that that hopefully will be in the next release. You can check if this is the case for you by running something like

```
MATCH (n:Person) WHERE n.birthYear IS NOT NULL AND toInteger(n.birthYear) IS NULL RETURN n.birthYear
```
(and likewise for deathYear)

If that query returns anything the database contains some malformed years that we aren't able to parse.

Best regards,
Pontus

**Comment by pontusmelke** [comment]
Hi,
we have identified and fixed the problem. The fix will be in the upcoming 5.15 release. It might still be worth to run the above query since it looks like you have some malformed year strings in your database which maybe is not desired?

Thanks for reporting. I will close this issue but don't hesitate to reopen or open a new one should the problem still occur in 5.15.

Best regards,
Pontus
